net/httpclient.go

```diff
@@ -0,0 +1,185 @@
+	MaxIdleConnsPerHost int
+	// MaxConnsPerHost see https://golang.org/pkg/net/http/#Transport.MaxConnsPerHost
+	MaxConnsPerHost int
+	// WriteBufferSize see https://golang.org/pkg/net/http/#Transport.WriteBufferSize
+	WriteBufferSize int
+	// ReadBufferSize see https://golang.org/pkg/net/http/#Transport.ReadBufferSize
+	ReadBufferSize int
+	// MaxResponseHeaderBytes see
+	// https://golang.org/pkg/net/http/#Transport.MaxResponseHeaderBytes
+	MaxResponseHeaderBytes int64
+	// Timeout sets all Timeouts, that are set to 0 to the given
+	// value. Basically it's the default timeout value.
+	Timeout time.Duration
+	// TLSHandshakeTimeout see
+	// https://golang.org/pkg/net/http/#Transport.TLSHandshakeTimeout,
+	// if not set or set to 0, its using Options.Timeout.
+	TLSHandshakeTimeout time.Duration
+	// IdleConnTimeout see
+	// https://golang.org/pkg/net/http/#Transport.IdleConnTimeout,
+	// if not set or set to 0, its using Options.Timeout.
+	IdleConnTimeout time.Duration
+	// ResponseHeaderTimeout see
+	// https://golang.org/pkg/net/http/#Transport.ResponseHeaderTimeout,
+	// if not set or set to 0, its using Options.Timeout.
+	ResponseHeaderTimeout time.Duration
+	// ExpectContinueTimeout see
+	// https://golang.org/pkg/net/http/#Transport.ExpectContinueTimeout,
+	// if not set or set to 0, its using Options.Timeout.
+	ExpectContinueTimeout time.Duration
+	// Tracer
+	Tracer opentracing.Tracer
+}
+
+type Transport struct {
+	tr     *http.Transport
+	tracer opentracing.Tracer
+}
+
+func NewHTTPRoundTripper(options Options, quit <-chan struct{}) *Transport {
+	// set default tracer
+	if options.Tracer == nil {
+		options.Tracer = &opentracing.NoopTracer{}
+	}
+
+	// set timeout defaults
+	if options.TLSHandshakeTimeout == 0 {
+		options.TLSHandshakeTimeout = options.Timeout
+	}
+	if options.IdleConnTimeout == 0 {
+		options.IdleConnTimeout = options.Timeout
+	}
+	if options.ResponseHeaderTimeout == 0 {
+		options.ResponseHeaderTimeout = options.Timeout
+	}
+	if options.ExpectContinueTimeout == 0 {
+		options.ExpectContinueTimeout = options.Timeout
+	}
+
+	htransport := &http.Transport{
+		DisableKeepAlives:      options.DisableKeepAlives,
+		DisableCompression:     options.DisableCompression,
+		ForceAttemptHTTP2:      options.ForceAttemptHTTP2,
+		MaxIdleConns:           options.MaxIdleConns,
+		MaxIdleConnsPerHost:    options.MaxIdleConnsPerHost,
+		MaxConnsPerHost:        options.MaxConnsPerHost,
+		WriteBufferSize:        options.WriteBufferSize,
+		ReadBufferSize:         options.ReadBufferSize,
+		MaxResponseHeaderBytes: options.MaxResponseHeaderBytes,
+		ResponseHeaderTimeout:  options.ResponseHeaderTimeout,
+		TLSHandshakeTimeout:    options.TLSHandshakeTimeout,
+		IdleConnTimeout:        options.IdleConnTimeout,
+		ExpectContinueTimeout:  options.ExpectContinueTimeout,
+	}
+
+	go func() {
+		for {
+			select {
+			case <-time.After(options.IdleConnTimeout):
+				htransport.CloseIdleConnections()
+			case <-quit:
+				return
+			}
+		}
```

Comment:
@szuecs Would make sense to catch panics in such goroutins?

Reply:
@miry In these kind of control loops, I don't think so. Why do you think there could be a panic?

Reply:
I checked the library, and found overuse of `panic`. 
Found one example:

```golang
func (pc *persistConn) closeLocked(err error) {
	if err == nil {
		panic("nil error")
```

Reply:
All filters are "expected" to panic since the idea was always to allow the community to contribute filters. Skipper itself handles eventual panics from filters on the [upper proxy layer](https://github.com/zalando/skipper/blob/master/proxy/proxy.go#L735).

Reply:
It's true that panics are recovered from in the proxy, but it's also true that this is a separate goroutine, so panics from would not be recovered from this way.

OTOH, the snippet above is not going to ever panic, because the function is always called with a non-nil error (https://github.com/golang/go/blob/master/src/net/http/transport.go#L772) I assume the panic is there only in case a test would call it in an invalid way.

There can be of course other causes for panics, like the system calls through the atomic package. We don't handle these, it's better to crash.